### PR TITLE
Expo and Sieve

### PR DESCRIPTION
**Pull request by azurerazor**
### Summary
Code for Binary Expo and Sieve of Eratosthenes

### Submission Link/Screenshots
https://vjudge.net/solution/54668344/2rrhUh1ih24habscIrab
https://codeforces.com/gym/102953/submission/284095772

note that the second submission link was done on gym and therefore is not visible unless you have solved the problem
It tests the sieve on 10^7 and it runs in less than 2 seconds at least

Also would like to know whether the const mod should be swapped to a function parameter

### Checklist
- [x] AC on at least 1 problem
- [x] Using tabs for indentation
- [x] Longest line at most 63 characters (with tab length 2)
- [x] Well documented

### CODE REVIEW

File: content/number-theory/Expo.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Author: Brian Grana
+ * Date: 2024-10-02
+ * Source: own
+ * Description: Given two integers b and p, computes b
+ * raised to the power of p under modulo.
+ * Time: $O(log(p))$
```
Comment (enwask) [inline content/number-theory/Expo.h]: `Time: $O(\log p)$`

File: content/number-theory/Expo.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * Status: tested on https://cses.fi/problemset/task/1095/
+ */
+#pragma once
+
+const long long mod = 1e9+7;
+long long expo(long long b, long long p) {
```
Comment (enwask) [inline content/number-theory/Expo.h]: assert nonnegative exponent? could avoid nasty bugs with the right shift

File: content/number-theory/Eratosthenes.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * Description: Prime sieve that generates a boolean vector where
+ * sieve[i] = 1 if and only if the number is prime.
+ * Time: $O(N\log\log (N))$
+ * Status: Tested on https://codeforces.com/gym/102953/problem/8
+ */
+#pragma once
+
+vector<bool> sieve(int n) {
+	vector<bool> sv(n+1, 1);
+	sv[0] = 0, sv[1] = 0;
```
Comment (enwask) [inline content/number-theory/Eratosthenes.h]: `sv[0] = sv[1] = 0;`

File: content/number-theory/Eratosthenes.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+	for (int i = 2; i * i <= n; i++) {
+		if (sv[i]) {
+			for (int j = i * i; j <= n; j += i) sv[j] = 0;
+		}
+	}
```
Comment (enwask) [inline content/number-theory/Eratosthenes.h:17]: do we want to compact this a little bit? i.e.
```cpp
for (int i = 2; i * i <= n; i++) if (sv[i]) {
    for (int j = i * i; j <= n; j += i) sv[j] = 0;
}
```

File: content/number-theory/Eratosthenes.h (hunk 3 above)
Comment (azurerazor) [reply content/number-theory/Eratosthenes.h]: fixed

File: content/number-theory/Eratosthenes.h (hunk 4 above)
Comment (azurerazor) [reply content/number-theory/Eratosthenes.h:17]: fixed

File: content/number-theory/Expo.h (hunk 1 above)
Comment (azurerazor) [reply content/number-theory/Expo.h]: fixed

File: content/number-theory/Expo.h (hunk 2 above)
Comment (azurerazor) [reply content/number-theory/Expo.h]: I don't think this is necessary personally but I'll see what other people think

File: content/number-theory/Expo.h
Comment (ssk4988) [inline content/number-theory/Expo.h:1]: no shot we're typing out long long every time

Comment (azurerazor) [reply content/number-theory/Expo.h:1]: that would be more of a general macros discussion

Comment (azurerazor) [reply content/number-theory/Expo.h:1]: fixed now

File: content/number-theory/Expo.h (hunk 2 above)
Comment (enwask) [reply content/number-theory/Expo.h]: did you see what other people think

File: content/number-theory/Expo.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Author: Brian Grana
+ * Date: 2024-10-02
+ * Source: own
+ * Description: Given two integers b and p, computes b
+ * raised to the power of p under modulo.
```
Comment (enwask) [inline content/number-theory/Expo.h]: `...computes $b^p$ under mod`

File: content/number-theory/Eratosthenes.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Author: Brian Grana
+ * Date: 2024-10-02
+ * Source: own
+ * Description: Prime sieve that generates a boolean vector where
+ * sieve[i] = 1 if and only if the number is prime.
```
Comment (enwask) [inline content/number-theory/Eratosthenes.h]: ...if and only if `i` is prime

File: content/number-theory/Expo.h (hunk 2 above)
Comment (azurerazor) [reply content/number-theory/Expo.h]: added comment // assert(p>0)

File: content/number-theory/Expo.h (hunk 5 above)
Comment (azurerazor) [reply content/number-theory/Expo.h]: fixed

File: content/number-theory/Eratosthenes.h (hunk 6 above)
Comment (azurerazor) [reply content/number-theory/Eratosthenes.h]: fixed

**Review by enwask — APPROVED** [review]

**Pull request merged**